File: oinkbot.js
const { Client, GatewayIntentBits, ActivityType, EmbedBuilder, REST, Routes } = require('discord.js');
const axios = require('axios');
const dotenv = require('dotenv');
const { exec } = require('child_process');
const util = require('util');
const execPromise = util.promisify(exec);

dotenv.config();

const llmapikey = process.env.API_KEY;
const llmapihost = process.env.API_URL;
const token = process.env.DISCORD_TOKEN;

const colors = {
    default: 0,
    teal: 0x1abc9c,
    dark_teal: 0x11806a,
    green: 0x2ecc71,
    dark_green: 0x1f8b4c,
    blue: 0x3498db,
    dark_blue: 0x206694,
    purple: 0x9b59b6,
    dark_purple: 0x71368a,
    magenta: 0xe91e63,
    dark_magenta: 0xad1457,
    gold: 0xf1c40f,
    dark_gold: 0xc27c0e,
    orange: 0xe67e22,
    dark_orange: 0xa84300,
    red: 0xe74c3c,
    dark_red: 0x992d22,
    lighter_grey: 0x95a5a6,
    dark_grey: 0x607d8b,
    light_grey: 0x979c9f,
    darker_grey: 0x546e7a,
    blurple: 0x7289da,
    greyple: 0x99aab5
};

const client = new Client({
    intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent
    ]
});

const sendRequest = async (data) => {
    const response = await axios.post(llmapihost, data, {
        headers: {
            'Content-Type': 'application/json',
            'Authorization': llmapikey
        },
        timeout: 100000
    });
        const responseData = response.data;
        const content = responseData.choices[0].message.content;
        const completionTokens = responseData.usage.completion_tokens;
        const totalTokens = responseData.usage.total_tokens;
        return { content, completionTokens, totalTokens };
};

const measureLatency = async (ipAddress, port) => {
    try {
        const start = Date.now();
        await execPromise(`nc -zv ${ipAddress} ${port}`);
        const latency = Date.now() - start;
        return latency;
    } catch (error) {
        console.error('Error measuring latency:', error);
        return null;
    }
};

client.once('ready', async () => {
    console.log(`We have logged in as ${client.user.tag}`);
    const rest = new REST({ version: '10' }).setToken(token);

    try {
        const commands = await rest.put(
            Routes.applicationCommands(client.user.id),
            { body: [
                {
                    name: 'ask',
                    description: 'Ask a question related to anything in the Verus Community!',
                    options: [
                        {
                            name: 'question',
                            type: 3, // STRING
                            description: 'The question you want to ask',
                            required: true
                        }
                    ]
                },
                {
                    name: 'ping',
                    description: 'Returns latency of the API and the bot!'
                }
            ] }
        );
        console.log(`Successfully registered ${commands.length} commands.`);
    } catch (error) {
        console.error('Error registering commands:', error);
    }

    changeStatus();
});

const changeStatus = async () => {
    const activities = [
        { name: 'with VerusCoins!', type: ActivityType.Playing },
        { name: 'for /ask command!', type: ActivityType.Watching },
        { name: 'Verus Community!', type: ActivityType.Competing },
        { name: '𝙊𝙞𝙣𝙠 papa!', type: ActivityType.Listening }
    ];
    let index = 0;
    setInterval(() => {
        client.user.setActivity(activities[index]);
        index = (index + 1) % activities.length;
    }, 5000);
};

client.on('interactionCreate', async interaction => {
    if (!interaction.isCommand()) return;

    if (interaction.commandName === 'ask') {
        await interaction.deferReply();
        const question = interaction.options.getString('question');

        try {
            const data = {
                "model": "lmstudio-community/VerusCommunity",
                "messages": [
                    {"role": "system", "content": "You are an intelligent assistant. You always provide well-reasoned answers that are both correct and helpful."},
                    {"role": "user", "content": question}
                ],
                "temperature": 0.05,
                "stream": false
            };
            const { content, completionTokens, totalTokens } = await sendRequest(data);

            const embed = new EmbedBuilder()
                .setTitle(question)
                .setDescription(content)
                .setColor(colors.green)
                .setFooter({ text: `Invoked by ${interaction.user.tag} | ${completionTokens} tokens out of ${totalTokens} total tokens generated.` });

            await interaction.editReply({ embeds: [embed] });
        } catch (error) {
            const embed = new EmbedBuilder()
                .setTitle('Boooo!!! Error')
                .setDescription(`Sorry, I'm having trouble understanding that. \`\`\`${error.message}\`\`\``)
                .setColor(colors.red);

            await interaction.editReply({ embeds: [embed] });
        }
    } else if (interaction.commandName === 'ping') {
        await interaction.deferReply();
        const latency = await measureLatency(llmapihost, llmapiport);
        const response = `Bot Latency: ${Math.round(client.ws.ping)}ms\nAPI Latency: ${latency ? `${latency}ms` : 'Error measuring latency'}`;
        await interaction.editReply(response);
    }
});

client.login(token);

File: oinkbot.py
import discord
from discord.ext import commands, tasks
from discord import app_commands
import httpx, asyncio, os
import socket
import timeit
from dotenv import load_dotenv
import datetime, requests

load_dotenv()

llmapikey = int(os.environ.get('API_KEY'))
llmapihost = str(os.environ.get('API_URL'))
token = str(os.environ.get('DISCORD_TOKEN'))

class colors:
    default = 0
    teal = 0x1abc9c
    dark_teal = 0x11806a
    green = 0x2ecc71
    dark_green = 0x1f8b4c
    blue = 0x3498db
    dark_blue = 0x206694
    purple = 0x9b59b6
    dark_purple = 0x71368a
    magenta = 0xe91e63
    dark_magenta = 0xad1457
    gold = 0xf1c40f
    dark_gold = 0xc27c0e
    orange = 0xe67e22
    dark_orange = 0xa84300
    red = 0xe74c3c
    dark_red = 0x992d22
    lighter_grey = 0x95a5a6
    dark_grey = 0x607d8b
    light_grey = 0x979c9f
    darker_grey = 0x546e7a
    blurple = 0x7289da
    greyple = 0x99aab5

def send_request(data):
    response = requests.request("POST", llmapihost, headers={"Content-Type": "application/json", "Authorization": llmapikey}, json=data)
    if response.status_code == 200:
        try:
            response_data = response.json()
            # Extract the required information
            content = response_data['choices'][0]['message']['content']
            completion_tokens = response_data['usage']['completion_tokens']
            total_tokens = response_data['usage']['total_tokens']
            # Print the extracted information
            print(f"Content: {content}")
            return content, completion_tokens, total_tokens
        except ValueError as e:
            return "Error: " + str(e)
    else:
        return "Error: " + str(response.status_code)

    
def measure_latency(ip_address, port):
    def connect_socket():
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(3)
        sock.connect((ip_address, port))
        sock.close()

    # Measure the time taken to establish a connection
    latency = timeit.timeit(connect_socket, number=1)
    latency_ms = latency * 1000
    return latency_ms

chatintents = discord.Intents.all()
global client
client = commands.Bot(command_prefix="oink ", help_command=None, intents=chatintents, owner_id=859368295823835136)

@tasks.loop(seconds=5)
async def changeStatus():
    await client.change_presence(status=discord.Status.online, activity=discord.Activity(type=discord.ActivityType.playing, name="with VerusCoins!"))
    await asyncio.sleep(3)
    await client.change_presence(status=discord.Status.online, activity=discord.Activity(type=discord.ActivityType.watching, name="for /ask command!"))
    await asyncio.sleep(3)
    await client.change_presence(status=discord.Status.online, activity=discord.Activity(type=discord.ActivityType.competing, name="Verus Community!"))
    await asyncio.sleep(3)
    await client.change_presence(status=discord.Status.online, activity=discord.Activity(type=discord.ActivityType.listening, name="𝙊𝙞𝙣𝙠 papa!"))
    await asyncio.sleep(3)

@client.event
async def on_ready():
    print(f'We have logged in as {client.user}')
    changeStatus.start()
    try:
        synced = await client.tree.sync()
        print(f"Successfully Synced {len(synced)} commands..")
    except Exception:
        print("Failed to sync commands")
    print("Enabled status updates..")

# @client.tree.error
# async def on_app_command_error(interaction: discord.Interaction, error):
#     embeddoerr = discord.Embed(title="Woahh!!", description="**Slow down** I can't keep pinging my APIs. Try again in {:.2f}s".format(error.retry_after), color=colors.red)
#     embeddoerr.set_footer(text=f"Command Error invoked by {interaction.user.name} on {datetime.datetime.now()}")
#     return await interaction.response.send_message(embed=embeddoerr, ephemeral=True)

@client.tree.command(name="ask", description="Ask a question related to anything in the Verus Community!")
@app_commands.guild_only()
async def askquestion(interaction: discord.Interaction, question: str):
    await interaction.response.defer()
    try:
        data = {
        "model": "lmstudio-community/VerusCommunity",
        "messages": [
            {"role": "system", "content": "You are an intelligent assistant. You always provide well-reasoned answers that are both correct and helpful."},
            {"role": "user", "content": question}
        ],
        "temperature": 0.05,
        "stream": False
        }
        answer, completion_tokens, total_tokens = await send_request(data)
        embed = discord.Embed(title=question, description=answer, color=colors.green)
        embed.set_footer(text=f"Invoked by {interaction.user.name} | {completion_tokens} tokens out of {total_tokens} total tokens generated.")
    except Exception as err:
        embed = discord.Embed(title="Boooo!!! Error", description=f"Sorry, I'm having trouble understanding that. ```{err}```", color=colors.red)
    await interaction.followup.send(embed=embed)
    
@client.tree.command(name="ping", description="Returns latency of the API and the bot!")
async def ping(interaction: discord.Interaction):
    await interaction.response.defer()
    response = f"Bot Latency: {round(client.latency * 1000)}ms"
    await interaction.followup.send(response, ephemeral=True)

client.run(token=token)